File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/GetAndSetStateTest.java
package dev.restate.sdk.blocking;

import static dev.restate.sdk.core.impl.ProtoUtils.greetingResponse;

import dev.restate.sdk.core.StateKey;
import dev.restate.sdk.core.TypeTag;
import dev.restate.sdk.core.impl.GetAndSetStateTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;

class GetAndSetStateTest extends GetAndSetStateTestSuite {

  private static class GetAndSetGreeter extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      String state = ctx.get(StateKey.of("STATE", TypeTag.STRING_UTF8)).get();

      ctx.set(StateKey.of("STATE", TypeTag.STRING_UTF8), request.getName());

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello " + state).build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService getAndSetGreeter() {
    return new GetAndSetGreeter();
  }

  private static class SetNullState extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      restateContext()
          .set(
              StateKey.of(
                  "STATE",
                  TypeTag.<String>using(
                      l -> {
                        throw new IllegalStateException("Unexpected call to serde fn");
                      },
                      l -> {
                        throw new IllegalStateException("Unexpected call to serde fn");
                      })),
              null);

      responseObserver.onNext(greetingResponse(""));
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService setNullState() {
    return new SetNullState();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/GetAndSetStateTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.StateKey
import dev.restate.sdk.core.TypeTag
import dev.restate.sdk.core.impl.GetAndSetStateTestSuite
import dev.restate.sdk.core.impl.testservices.GreeterGrpcKt
import dev.restate.sdk.core.impl.testservices.GreetingRequest
import dev.restate.sdk.core.impl.testservices.GreetingResponse
import dev.restate.sdk.core.impl.testservices.greetingResponse
import io.grpc.BindableService
import kotlinx.coroutines.Dispatchers

internal class GetAndSetStateTest : GetAndSetStateTestSuite() {
  private class GetAndSetGreeter :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()

      val state = ctx.get(StateKey.of("STATE", TypeTag.STRING_UTF8))!!
      ctx.set(StateKey.of("STATE", TypeTag.STRING_UTF8), request.getName())

      return greetingResponse { message = "Hello $state" }
    }
  }

  override fun getAndSetGreeter(): BindableService {
    return GetAndSetGreeter()
  }

  override fun setNullState(): BindableService {
    throw UnsupportedOperationException("The kotlin type system enforces non null state values")
  }
}
